src/graph/validation/validate_graph.rs

```diff
@@ -0,0 +1,243 @@
+// License: MIT
+// Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+//! Methods for validating the acyclicity and connectedness of a
+//! [`ComponentGraph`].
+
+use std::collections::BTreeSet;
+
+use crate::{Edge, Error, Node};
+
+use super::ComponentGraphValidator;
+
+impl<N, E> ComponentGraphValidator<'_, N, E>
+where
+    N: Node,
+    E: Edge,
+{
+    pub(super) fn validate_connected_graph(&self, root: &N) -> Result<(), Error> {
+        let root_id = root.component_id();
+        let mut visited = BTreeSet::new();
+        let mut queue = vec![root_id];
+        visited.insert(root_id);
+        while let Some(node_id) = queue.pop() {
+            for successor in self.cg.successors(node_id)? {
+                visited.insert(successor.component_id());
+                queue.push(successor.component_id());
+            }
+        }
+
+        let unvisited = self
+            .cg
+            .components()
+            .map(|n| n.component_id())
+            .filter(|id| !visited.contains(id))
+            .collect::<Vec<_>>();
+
+        if !unvisited.is_empty() {
+            return Err(Error::invalid_graph(format!(
+                "Nodes {:?} are not connected to the root.",
+                unvisited
+            )));
+        }
+
+        Ok(())
+    }
+
+    pub(super) fn validate_acyclicity(
+        &self,
+        node: &N,
+        mut predecessors: Vec<u64>,
+    ) -> Result<(), Error> {
+        predecessors.push(node.component_id());
```

Comment:
Doesn't petgraph provide a check for this already? https://docs.rs/petgraph/latest/petgraph/algo/fn.is_cyclic_directed.html

If it is not suitable for us for some reason maybe add a comment to clarify as maybe other people try to use it?

Reply:
Their implementation doesn't show where the cycle is, this implementation gives us better logs.